LibCompress+Utilities: Add LZMA

timschumi opened the pull request:
Pain.

This is annoyingly close to the reference implementation, at the very least the math-heavy parts that actually get into decoding instead of the high-level logic (granted, not like one can diverge much anyways if that's the only real source of "that is the formula you need to use" and "you need that many different probability counters and use them in a specific order"). The official specification is unfortunately basically just the reference implementation with a bit of text around the logic parts, and I wasn't able to find any human-language specifications that have actually been completed.

I hope that this is still acceptable, considering that the specification/reference implementation has been placed in the public domain.

## Remaining TODOs/discussion points

 - ~~Fuzzer: Not viable at the moment, since the archive can legitimately request up to 4GB of lookback buffer size, which instantly OOMs LibFuzzer.~~
 - ~~Specification quotes: The implementation currently features a mix of quotes from the specification and some comments of my own. Suggestions on how to make those visually distinct would be very welcome, if that is a priority.~~
 - ~~Tests: Will add tomorrow.~~
 - ~~Rough edges: Nothing is explicitly unimplemented at the moment, but there are some noticeable edge cases that I haven't looked into yet (for example, some of the testing files from the SDK don't even load properly).~~

kleinesfilmroellchen commented on Userland/Libraries/LibCompress/Lzma.h:
```diff
@@ -0,0 +1,124 @@
+
+class LzmaDecompressor : public Stream {
+public:
+    static ErrorOr<NonnullOwnPtr<LzmaDecompressor>> create(NonnullOwnPtr<Stream>, LzmaDecompressorOptions const&);
```
Should this not take `Handle<Stream>`?

timschumi replied:
It's called `MaybeOwned` now, but yes, it should (at the very least for LZMA2). :^)

kleinesfilmroellchen commented on Userland/Libraries/LibArchive/Lzma.h:
```diff
@@ -0,0 +1,43 @@
+
+class Lzma {
+public:
+    static ErrorOr<NonnullOwnPtr<Compress::LzmaDecompressor>> create(NonnullOwnPtr<Stream>);
```
Same here with Handle

kleinesfilmroellchen commented on Userland/Utilities/lzcat.cpp:
```diff
@@ -0,0 +1,31 @@
+    StringView filename;
+
+    Core::ArgsParser args_parser;
+    args_parser.add_positional_argument(filename, "File to decompress", "file");
```
Should probably have a short help text like "Decompress and print an LZMA archive"

kleinesfilmroellchen commented on Userland/Utilities/lzcat.cpp:
```diff
@@ -0,0 +1,31 @@
+    NonnullOwnPtr<Stream> stream = TRY(Core::File::open(filename, Core::File::OpenMode::Read));
+    stream = TRY(Archive::Lzma::create(move(stream)));
+
+    // Arbitrarily chosen buffer size.
+    Array<u8, 4096> buffer;
```
Use PAGE_SIZE?

timschumi replied:
`PAGE_SIZE` is not consteval on Lagom, so we can't use it here. It also isn't nearly as important to align to pages compared to heap allocations. In this case, it really is just a nice and round number.

kleinesfilmroellchen commented on Meta/Lagom/Fuzzers/FuzzLzmaDecompression.cpp:
```diff
@@ -0,0 +1,34 @@
+extern "C" int LLVMFuzzerTestOneInput(uint8_t const* data, size_t size)
+{
+    // These settings have been taken from the a.lzma example in the LZMA specification.
+    // Fuzzing the settings is not currently viable as archives can request a dictionary size of up to 4GB.
```
Can we fuzz all options *except* the dictionary size? i.e. overwrite the dictionary size option after reading the options from the provided data.

timschumi replied:
We could do that, but that would result in files that look like LZMA archives (not that they have some kind of magic), but would have to be changed first to yield the same results as the fuzzer.

What we _could_ do is to define some arbitrary "largest reasonable dictionary size" and reject any test inputs that have a larger dictionary size than that. LibFuzzer has a dedicated return value to indicate "don't add this to the test corpus", which would be perfect for this.

kleinesfilmroellchen replied:
That sounds like the ideal solution to me then.

kleinesfilmroellchen commented on Meta/Lagom/Fuzzers/FuzzLzmaDecompression.cpp:
```diff
@@ -0,0 +1,34 @@
+        return 0;
+    auto decompressor = decompressor_or_error.release_value();
+    while (!decompressor->is_eof()) {
+        auto maybe_error = decompressor->discard(4096);
```
PAGE_SIZE as well probably

timschumi commented:
Renamed `Archive::Lzma` to `Archive::LzmaDecompressor`, to match `Compress::LzmaDecompressor`. But now this should really be ready for merging. :^)

GMTA commented:
> Specification quotes: The implementation currently features a mix of quotes from the specification and some comments of my own. Suggestions on how to make those visually distinct would be very welcome, if that is a priority.

I try to explicitly reference section numbers when possible, and literally quote (`"The spec says that..."`) the parts that I copy over from the spec.

timschumi commented:
I'm afraid the specification is a bit short on section numbers, and it uses section names/headers _very_ sparingly. I will try adding quotes around the quotes and prefixing them with some kind of distinctive text to see how it looks in practice.

timschumi commented:
Changes in latest push:
 - Rebased onto and adapted to the latest Stream function name changes.
 - Removed some unnecessary braces around one line for-loops.
 - Visually marked verbatim specification comments by wrapping them in `"`.
 - Replaced the specification vs. reference implementation FIXME with a "Note", since I experimentally confirmed that the reference implementation is most likely correct.

IdanHo commented on Userland/Libraries/LibCompress/Lzma.cpp:
```diff
@@ -0,0 +1,604 @@
+        m_range_decoder_range = bound;
+        TRY(normalize_range_decoder());
+        return 0;
+    } else {
```
no need for else after return

IdanHo replied:
nitpicks can be ignored when reasonable :slightly_smiling_face:

IdanHo commented on Userland/Libraries/LibCompress/Lzma.cpp:
```diff
@@ -0,0 +1,604 @@
+    u8 previous_byte = 0;
+    if (m_total_decoded_bytes > 0) {
+        auto read_bytes = MUST(m_output_buffer.read_with_seekback({ &previous_byte, sizeof(previous_byte) }, 1));
+        VERIFY(read_bytes.size() == sizeof(previous_byte));
```
Not sure I understand this verify, if `read_with_seekback` didn't return an error `read_bytes` is a guarenteed to return 1 byte right? Otherwise it's a bug in the `read_with_seekback` implementation? If so, this seems a bit excessive, we should either assert this inside the implementation itself, or not at all.

IdanHo replied:
(same for everywhere else below)

IdanHo commented on Userland/Libraries/LibCompress/Lzma.cpp:
```diff
@@ -0,0 +1,604 @@
+    u16 literal_state_bits_from_output = previous_byte >> (8 - m_options.literal_context_bits);
+    u16 literal_state = literal_state_bits_from_position << m_options.literal_context_bits | literal_state_bits_from_output;
+
+    Span<Probability> selected_probability_table = m_literal_probabilities.span().slice(0x300 * literal_state, 0x300);
```
We use this `0x300` constant in multiple places, maybe make it a constexpr constant?

IdanHo replied:
I think `probability_table_size` is fine

IdanHo commented on Userland/Libraries/LibCompress/Lzma.cpp:
```diff
@@ -0,0 +1,604 @@
+}
+
+LzmaDecompressor::LzmaLengthDecoderState::LzmaLengthDecoderState()
+{
```
ditto regarding the helper from above

IdanHo commented on Userland/Libraries/LibCompress/Lzma.cpp:
```diff
@@ -0,0 +1,604 @@
+    if (m_output_buffer.used_space() > 0)
+        return false;
+
+    if (m_options.uncompressed_size.has_value()) {
+        return m_total_decoded_bytes >= m_options.uncompressed_size.value();
+    }
```
nitpick, unnecessary curlies

ADKaster commented on Userland/Libraries/LibArchive/Lzma.cpp:
```diff
@@ -0,0 +1,96 @@
+
+    literal_context_bits = input_bits % 9;
+    input_bits /= 9;
+    VERIFY(literal_context_bits >= 0 && literal_context_bits <= 8);
```
Just to confirm: These properties on the bits are mandated by either the math here in this function or by the caller of this function? (i.e. we won't see any fuzzer crashes hitting these asserts, right?)

ADKaster commented on Userland/Libraries/LibArchive/Lzma.h:
```diff
@@ -0,0 +1,45 @@
+
+class LzmaDecompressor {
+public:
+    static ErrorOr<NonnullOwnPtr<Compress::LzmaDecompressor>> create(MaybeOwned<Stream>);
```
Archive::LzmaDecompressor::create() returns a Compress::LzmaDecompressor? This seems kind of odd, but then I didn't cross reference it with the other Decompressors that mostly just redirect to LibCompress.

timschumi replied:
The reason why I made this separate is because LZMA technically defines an 'archive' format (which is more of a stream container, really).

It is a really puny archive format, because all it does is prefixing it with a small header that notes down the decoder settings for the contained stream, but it is distinct from the other LibCompress-supported algorithms that really only dump the raw stream into a file and call it a day.

The reason why it returns a `Compress::LzmaDecompressor` is because we read the header upon creation of the stream, configure the underlying `LibCompress` stream once, and would then just delegate all the time anyways. Returning the underlying Stream directly and not implementing the whole Stream interface for Archive::LzmaDecompressor again seemed like a better idea.

Would we rather want to have a `Compress::LzmaDecompressor::create_from_archive` (/container?) and `Compress::LzmaDecompressor::create_from_raw_stream`, where one relies on the other?

timschumi replied:
From Idans comment about the same topic:

> The distiniction between a compression format and an archival format, is that a compression format takes a stream of bytes and compresses/decompresses it, while an archive format packs together one or more files, with the file contents possibly being compressed using a compression format.

Ok, in that case LZMA 'archives' wouldn't be an actual archive because (in comparison to an LZMA compressed stream) they only provide the settings for the singular stream that they contain.

However, looking at what's to come, are LZMA2 and XZ archive formats then? Should they also rather be in LibCompress?

Neither of them have a concept of one or multiple files, both just encode a stream of data, but this time in multiple ways. LZMA2 splits the stream in parts and either stores them LZMA-compressed or uncompressed. XZ allows compressing its data stream using LZMA2 and/or a bunch of of other so-called filters. None of them know what 'files' are, if a user needs that they have to TAR it up first.

IdanHo replied:
Comparing to the Gzip, which we already have implemented in serenity, it seems to match up: It's based on Deflate, it adds a short header at the start with the compression options, and is possibly used to compress tar archives.
Currently the implementation for Gzip is in LibCompress, matching my expectations at least.

The only unfortunate part about LZMA is that the name is used both for the raw stream and the stream with the header part.

I would say all 3 of LZMA, LZMA2 and XZ should end up in LibCompress.

ADKaster commented on Userland/Utilities/lzcat.cpp:
```diff
@@ -0,0 +1,32 @@
+    args_parser.add_positional_argument(filename, "File to decompress", "file");
+    args_parser.parse(arguments);
+
+    NonnullOwnPtr<Stream> stream = TRY(Core::File::open(filename, Core::File::OpenMode::Read));
```
Thoughts on using that "open file or standard stream" thing here? or does Posix/Linux lzcat not support `-` as an arg.

IdanHo commented on Userland/Libraries/LibCompress/Lzma.cpp:
```diff
@@ -0,0 +1,604 @@
+            if (m_state < 4)
+                m_state = 0;
+            else if (m_state < 10)
+                m_state = m_state - 3;
```
nitpick, but `-=`

ADKaster reviewed (commented):
Couple thoughts for clarification, but looks good :^)

IdanHo commented on Userland/Libraries/LibCompress/Lzma.cpp:
```diff
@@ -0,0 +1,604 @@
+            // "At first the LZMA decoder must check that it doesn't exceed
+            //  specified uncompressed size."
+            if (m_options.uncompressed_size.has_value() && m_options.uncompressed_size.value() <= m_total_decoded_bytes)
+                return Error::from_string_literal("Exceeded uncompressed archive size while decoding literal symbol");
```
we already check this above, so this is unrechable. It would only be reachable if we did another iteration of the loop to check for the EOF marker, which we don't.

IdanHo replied:
(I would keep the spec text, but add a comment explaining it's not unecessary, assuming you agree with me of course :slightly_smiling_face:)

IdanHo commented on Userland/Libraries/LibCompress/Lzma.cpp:
```diff
@@ -0,0 +1,604 @@
+            //  "End of stream" marker, so we can stop decoding and check finishing
+            //  condition in Range Decoder"
+            if (m_rep0 == 0xFFFFFFFF) {
+                if (m_range_decoder_code != 0)
+                    return Error::from_string_literal("Found 'end of stream' marker, but range decoder is in unfinished state");
+
+                m_found_end_of_stream_marker = true;
+                continue;
```
We set `m_found_end_of_stream_marker` (this is the only place where we do so), and immediately continue the loop, at which point we break out of it using the loop end condition at the top. We then go into the post-loop check for the code validity if we reached the end of stream marker.
AKA: This check for the final code value is duplicated, we should remove one of the copies.

IdanHo commented on Userland/Libraries/LibCompress/Lzma.cpp:
```diff
@@ -0,0 +1,604 @@
+                continue;
+            }
+
+            // "If uncompressed size is defined, LZMA decoder must check that it doesn't
+            //  exceed that specified uncompressed size."
+            if (m_options.uncompressed_size.has_value() && m_options.uncompressed_size.value() <= m_total_decoded_bytes)
+                return Error::from_string_literal("Exceeded uncompressed archive size while decoding simple match symbol");
```
same as above

IdanHo reviewed (commented):
nice work, mostly nitpicks

timschumi commented on Userland/Libraries/LibArchive/Lzma.h:
```diff
@@ -0,0 +1,45 @@
+};
+static_assert(sizeof(LzmaHeader) == 13);
+
+class LzmaDecompressor {
+public:
+    static ErrorOr<NonnullOwnPtr<Compress::LzmaDecompressor>> create(MaybeOwned<Stream>);
+};
```
The question regarding the return type I have started to discuss in a response to Andrew, so we should probably continue to do that there exclusively.

timschumi commented on Tests/LibCompress/TestLzma.cpp:
```diff
@@ -0,0 +1,214 @@
+
+    auto stream = MUST(try_make<FixedMemoryStream>(compressed));
+    auto decompressor = MUST(Compress::LzmaDecompressor::create(move(stream), options));
+    // FIXME: This should detect that there is still remaining (non-'end of stream') data after the expected data and reject the last read.
+    //        As of now, this test accepts any result, except for crashes.
+    (void)decompressor->read_until_eof(PAGE_SIZE);
```
> Are we going to fix this FIXME?

I plan to, yes. However, since my first attempt at this led to a bunch of erroneous failures and end-of-stream detection will have a significant effect on formats that embed LZMA, I decided to let it be for now and revisit it later, focusing on getting the main parts of the decoder (and the followup support for LZMA2 and XZ) finished up and merged first.

> Also, we might want to expose a is_corrupted helper that returns true if we reached the 'known' uncompressed size but m_range_decoder_code is non-0? The spec says that's a good indicator for an incorrectly decoded stream (and you already check that when we reach the explicit end indicator), so we can assert that returns true here.

The specification indeed mentions that this is a good way of checking for corruption, but it unfortunately doesn't seem to hold up if the archive contains an uncompressed size _and_ an end of stream marker (since it still has to read the end marker at that point). This is also what I suspect was causing my failures that shouldn't have been there.

In any case, the current mode of operation of this implementation is to stop and report an error even for things that would otherwise just be "corrupted", so I'd rather have it report that immediately if it even just notices that something is wrong.

IdanHo replied:
I see, let's leave that to a future PR then :+1:

timschumi reviewed (commented):
Fixed a large part of the nitpicks, left comments on the others.

IdanHo commented on Userland/Libraries/LibCompress/Lzma.h:
```diff
@@ -0,0 +1,126 @@
+        Probability m_second_choice_probability { default_probability };
+
+        static constexpr size_t maximum_number_of_position_bits = 4;
+        Array<Array<Probability, (1 << 3)>, (1 << maximum_number_of_position_bits)> m_low_length_probabilities;
+        Array<Array<Probability, (1 << 3)>, (1 << maximum_number_of_position_bits)> m_medium_length_probabilities;
+        Array<Probability, (1 << 8)> m_high_length_probabilities;
```
Ah, if it's just the chosen value per tree then it's fine, I thought `1 << 3` was being duplicated, but it sounds like it's actually 2 different values, low tree length, and medium tree length, which just ended up as the same value.

IdanHo commented on Userland/Libraries/LibCompress/Lzma.cpp:
```diff
@@ -0,0 +1,604 @@
+    m_range_decoder_range >>= 1;
+    m_range_decoder_code -= m_range_decoder_range;
+
+    // The specification (or rather, the reference implementation) does not specify what "t" is. As such, the name has been carried over unchanged.
+    // Since the human language description on Wikipedia decides the return value by comparing range and code, "t" appears to be a simple helper for a branchless implementation.
+    u32 t = 0 - (m_range_decoder_code >> 31);
```
Yep, I prefer so, simply to make the code a tiny bit more readable.

IdanHo reviewed (approved):
LGTM, thanks! Support for new compression formats is always exciting :^)
